File: src/crawler/extract.js
/** HTML extraction for crawled pages (app/crawler/extract.py), using Cheerio on htmlparser2. */

import { createHash } from 'node:crypto';
import { PY_SPACE_CHARS, pyLower, pySplit, pyStrip } from '../utils/pytext.js';
import { normalizeCrawlUrl } from './urls.js';
import { detectHtmlRedirect, isNextErrorShell } from './html-redirects.js';
import {
  attrValue,
  decodeContents,
  decodeDocument,
  descendants,
  getText,
  isAttached,
  isElement,
  loadHtml,
  removeNode,
} from './dom.js';

// Removed from the stored main content: chrome, scripts and non-content widgets.
const STRIP_TAGS = new Set([
  'script', 'style', 'noscript', 'template', 'svg', 'iframe', 'canvas', 'form',
  'nav', 'header', 'footer', 'aside', 'dialog', 'button', 'select', 'input', 'textarea',
  'link', 'meta',
]);
const STRIP_ROLES = new Set(['navigation', 'banner', 'contentinfo', 'search', 'dialog', 'alert']);
const S = PY_SPACE_CHARS; // Python `\s`
const STRIP_HINT = new RegExp(
  `(^|[${S}_-])(cookie|consent|gdpr|newsletter|breadcrumb|skip-link|sr-only|modal|popup)([${S}_-]|$)`,
  'iu',
);
const WS_RUN = new RegExp(`[${S}]+`, 'gu');
const ROBOTS_SPLIT = new RegExp(`[${S},]+`, 'u');

/** `_text`: whitespace-collapsed get_text or null. */
function text(node) {
  if (!node) return null;
  const value = pyStrip(getText(node).replace(WS_RUN, ' '));
  return value || null;
}

/** `_attr`: stripped attribute value or null. */
function attr(node, name) {
  if (!node) return null;
  const value = attrValue(node, name);
  return typeof value === 'string' && pyStrip(value) ? pyStrip(value) : null;
}

function meta($, name) {
  const target = pyLower(name);
  const node = $('meta')
    .toArray()
    .find((m) => typeof m.attribs?.name === 'string' && pyLower(m.attribs.name) === target);
  return attr(node, 'content');
}

function hrefs(nodes, baseUrl) {
  const urls = [];
  for (const anchor of nodes) {
    if (anchor.name !== 'a' || anchor.attribs?.href === undefined) continue;
    const href = attr(anchor, 'href');
    if (href === null || ['#', 'mailto:', 'tel:', 'javascript:', 'data:'].some((p) => href.startsWith(p))) continue;
    const rel = pyLower(attr(anchor, 'rel') ?? '');
    if (pySplit(rel).includes('nofollow')) continue;
    const normalised = normalizeCrawlUrl(href, baseUrl);
    if (normalised) urls.push(normalised);
  }
  return [...new Set(urls)];
}

function mainRoot($) {
  for (const selector of ['main', '[role=main]', 'article']) {
    const candidates = $(selector).toArray();
    if (candidates.length) {
      // The largest candidate is the page body (some layouts nest several <article>s).
      let best = candidates[0];
      let bestLen = [...getText(best)].length;
      for (const c of candidates.slice(1)) {
        const len = [...getText(c)].length;
        if (len > bestLen) {
          best = c;
          bestLen = len;
        }
      }
      return best;
    }
  }
  return $('body').get(0) ?? null;
}

function stripChrome(root) {
  for (const node of [...descendants(root)]) {
    if (STRIP_TAGS.has(node.name) && isAttached(node, root)) removeNode(node);
  }
  for (const node of [...descendants(root)]) {
    if (!isAttached(node, root)) continue;
    const role = pyLower(attr(node, 'role') ?? '');
    const hints = `${attr(node, 'id') ?? ''} ${attr(node, 'class') ?? ''}`;
    const hidden = node.attribs?.hidden !== undefined || (attr(node, 'aria-hidden') ?? '') === 'true';
    if (STRIP_ROLES.has(role) || hidden || STRIP_HINT.test(hints)) removeNode(node);
  }
}

/** True when a 2xx HTML page has effectively no usable content (see Python `is_unusable`). */
export function isUnusable({ title, h1, contentText, errorShell = false }) {
  const hasText = Boolean(pyStrip(contentText));
  if (errorShell) return !hasText;
  return !hasText && !pyStrip(title ?? '') && !pyStrip(h1 ?? '');
}

export function extractPage(html, pageUrl, { xRobotsTag = null } = {}) {
  const $ = loadHtml(html);
  const doc = $.root().get(0);
  for (const node of [...allNodes(doc)]) if (node.type === 'comment') removeNode(node);

  const title = text($('title').get(0));
  const canonicalNode = $('link')
    .toArray()
    .find((l) => typeof l.attribs?.rel === 'string' && pyLower(l.attribs.rel).includes('canonical'));
  const canonicalHref = attr(canonicalNode, 'href');
  const canonical = canonicalHref ? normalizeCrawlUrl(canonicalHref, pageUrl) : null;
  const language = attr($('html').get(0), 'lang');
  const robotsDirectives = new Set(
    pyLower([meta($, 'robots'), xRobotsTag].filter(Boolean).join(' ')).split(ROBOTS_SPLIT),
  );
  const keywordsMeta = meta($, 'keywords');
  const allLinks = hrefs([...descendants(doc)], pageUrl);
  const htmlRedirect = detectHtmlRedirect(html, $);
  const errorShell = isNextErrorShell(html);

  const root = mainRoot($);
  const h1 = text($('h1').get(0));
  const rootNode = root ?? doc;
  stripChrome(rootNode);
  const headings = [...descendants(rootNode)]
    .filter((n) => ['h1', 'h2', 'h3'].includes(n.name))
    .map((n) => [n.name, text(n)])
    .filter(([, t]) => t);
  const contentText = text(rootNode) ?? '';
  const contentHtml = root ? pyStrip(decodeContents(root)) : decodeDocument(doc);
  return {
    title,
    h1,
    metaDescription: meta($, 'description'),
    canonicalUrl: canonical,
    language,
    noindex: robotsDirectives.has('noindex') || robotsDirectives.has('none'),
    headings,
    contentHtml,
    contentText,
    contentHash: createHash('sha256').update(contentText, 'utf8').digest('hex'),
    keywords: (keywordsMeta ?? '').split(',').map((k) => pyStrip(k)).filter(Boolean).slice(0, 50),
    links: allLinks,
    contentLinks: hrefs([...descendants(rootNode)], pageUrl),
    htmlRedirect,
    isErrorShell: errorShell,
    isEmpty: isUnusable({ title, h1, contentText, errorShell }),
  };
}

function* allNodes(node) {
  for (const child of node.children ?? []) {
    yield child;
    if (isElement(child) || child.type === 'root') yield* allNodes(child);
  }
}

File: app/crawler/extract.py
"""HTML extraction for crawled pages (BeautifulSoup, stdlib `html.parser` backend)."""

from __future__ import annotations

import hashlib
import re
from dataclasses import dataclass, field

from bs4 import BeautifulSoup, Comment, Tag

from app.crawler.html_redirects import HtmlRedirect, detect_html_redirect, is_next_error_shell
from app.crawler.urls import normalize_crawl_url

# Removed from the stored main content: chrome, scripts and non-content widgets.
_STRIP_TAGS = (
    "script", "style", "noscript", "template", "svg", "iframe", "canvas", "form",
    "nav", "header", "footer", "aside", "dialog", "button", "select", "input", "textarea",
    "link", "meta",
)  # fmt: skip
_STRIP_ROLES = {"navigation", "banner", "contentinfo", "search", "dialog", "alert"}
_STRIP_HINT = re.compile(
    r"(^|[\s_-])(cookie|consent|gdpr|newsletter|breadcrumb|skip-link|sr-only|modal|popup)"
    r"([\s_-]|$)",
    re.IGNORECASE,
)
_WS = re.compile(r"\s+")


@dataclass
class ExtractedPage:
    title: str | None
    h1: str | None
    meta_description: str | None
    canonical_url: str | None
    language: str | None
    noindex: bool
    headings: list[tuple[str, str]]  # (tag, text) from the main content
    content_html: str
    content_text: str
    content_hash: str
    keywords: list[str] = field(default_factory=list)
    links: list[str] = field(default_factory=list)  # every <a href> in the document (absolute)
    content_links: list[str] = field(default_factory=list)  # <a href> inside main content
    # Redirect declared inside the HTML (Next.js NEXT_REDIRECT digest or meta refresh).
    # Raw and unvalidated: callers must normalise + scope/SSRF-check before following it.
    html_redirect: HtmlRedirect | None = None
    is_error_shell: bool = False  # Next.js `<html id="__next_error__">` document
    # No usable page content (see `is_unusable`); such pages are not normal content pages.
    is_empty: bool = False


def _text(node: Tag | None) -> str | None:
    if node is None:
        return None
    value = _WS.sub(" ", node.get_text(" ", strip=True)).strip()
    return value or None


def _attr(node: Tag | None, name: str) -> str | None:
    if node is None:
        return None
    value = node.get(name)
    if isinstance(value, list):
        value = " ".join(value)
    return value.strip() if isinstance(value, str) and value.strip() else None


def _meta(soup: BeautifulSoup, name: str) -> str | None:
    node = soup.find("meta", attrs={"name": re.compile(f"^{re.escape(name)}$", re.I)})
    return _attr(node if isinstance(node, Tag) else None, "content")


def _hrefs(root: Tag, base_url: str) -> list[str]:
    urls: list[str] = []
    for anchor in root.find_all("a", href=True):
        if not isinstance(anchor, Tag):
            continue
        href = _attr(anchor, "href")
        if href is None or href.startswith(("#", "mailto:", "tel:", "javascript:", "data:")):
            continue
        rel = (_attr(anchor, "rel") or "").lower()
        if "nofollow" in rel.split():
            continue
        normalised = normalize_crawl_url(href, base_url)
        if normalised:
            urls.append(normalised)
    return list(dict.fromkeys(urls))


def _main_root(soup: BeautifulSoup) -> Tag:
    for selector in ("main", "[role=main]", "article"):
        candidates = [c for c in soup.select(selector) if isinstance(c, Tag)]
        if candidates:
            # The largest candidate is the page body (some layouts nest several <article>s).
            return max(candidates, key=lambda c: len(c.get_text(" ", strip=True)))
    return soup.body if isinstance(soup.body, Tag) else soup


def _strip_chrome(root: Tag) -> None:
    for node in root.find_all(_STRIP_TAGS):
        node.decompose()
    for node in root.find_all(True):
        if not isinstance(node, Tag) or node.decomposed:
            continue
        role = (_attr(node, "role") or "").lower()
        hints = f"{_attr(node, 'id') or ''} {_attr(node, 'class') or ''}"
        hidden = node.has_attr("hidden") or (_attr(node, "aria-hidden") or "") == "true"
        if role in _STRIP_ROLES or hidden or _STRIP_HINT.search(hints):
            node.decompose()


def extract_page(html: str, page_url: str, *, x_robots_tag: str | None = None) -> ExtractedPage:
    soup = BeautifulSoup(html, "html.parser")
    for comment in soup.find_all(string=lambda s: isinstance(s, Comment)):
        comment.extract()

    title = _text(soup.title if isinstance(soup.title, Tag) else None)
    canonical_node = soup.find("link", rel=lambda v: bool(v) and "canonical" in str(v).lower())
    canonical_href = _attr(canonical_node if isinstance(canonical_node, Tag) else None, "href")
    canonical = normalize_crawl_url(canonical_href, page_url) if canonical_href else None
    html_node = soup.find("html")
    language = _attr(html_node if isinstance(html_node, Tag) else None, "lang")
    robots_directives = set(
        re.split(r"[\s,]+", " ".join(filter(None, [_meta(soup, "robots"), x_robots_tag])).lower())
    )
    keywords_meta = _meta(soup, "keywords")
    all_links = _hrefs(soup, page_url)
    html_redirect = detect_html_redirect(html, soup)
    error_shell = is_next_error_shell(html)

    root = _main_root(soup)
    first_h1 = soup.find("h1")
    h1 = _text(first_h1 if isinstance(first_h1, Tag) else None)
    _strip_chrome(root)
    headings = [
        (node.name, text)
        for node in root.find_all(["h1", "h2", "h3"])
        if isinstance(node, Tag) and (text := _text(node))
    ]
    content_text = _text(root) or ""
    content_html = root.decode_contents().strip() if root is not soup else str(root)
    return ExtractedPage(
        title=title,
        h1=h1,
        meta_description=_meta(soup, "description"),
        canonical_url=canonical,
        language=language,
        noindex=bool(robots_directives & {"noindex", "none"}),
        headings=headings,
        content_html=content_html,
        content_text=content_text,
        content_hash=hashlib.sha256(content_text.encode("utf-8")).hexdigest(),
        keywords=[k.strip() for k in (keywords_meta or "").split(",") if k.strip()][:50],
        links=all_links,
        content_links=_hrefs(root, page_url),
        html_redirect=html_redirect,
        is_error_shell=error_shell,
        is_empty=is_unusable(
            title=title, h1=h1, content_text=content_text, error_shell=error_shell
        ),
    )


def is_unusable(
    *, title: str | None, h1: str | None, content_text: str, error_shell: bool = False
) -> bool:
    """True when a 2xx HTML page has effectively no usable content.

    A page missing only one of title / H1 / canonical is still valid as long as it has
    something to link to or from; only a page with no title, no H1 and no visible content
    (or a Next.js error shell without visible content) is unusable. Canonical alone never
    makes a page usable.
    """
    has_text = bool(content_text.strip())
    if error_shell:
        return not has_text
    return not has_text and not (title or "").strip() and not (h1 or "").strip()
